Sources/ProjectSpec/Encoding.swift

```diff
@@ -0,0 +1,19 @@
+import Foundation
+import JSONUtilities
+
+public protocol JSONDictionaryEncodable {
+    func toJSONDictionary() -> JSONDictionary
+}
+
+public protocol JSONArrayEncodable {
+    func toJSONArray() -> JSONArray
+}
+
+public protocol JSONPrimitiveEncodable {
+    func toJSONValue() -> JSONRawType
+}
+
+public protocol JSONDynamicEncodable {
+    // returns JSONDictionary or JSONArray or JSONRawType
+    func toJSONValue() -> Any
```

Comment:
I'm not sure we need all these protocols as they aren't really used on their own. I think all the functions could just be the `func toJSONValue() -> Any`. Maybe just the top level spec would have a function that returned a dictionary.